TicTacToe/App.fs

```diff
@@ -4,172 +4,170 @@ open Elmish
+        match oWins with         
+        | Some p -> p 
+        | _ -> 
+
+        match anyMoreMoves model with
+        | true -> StillPlaying
+        | false -> Draw
+
+
+    /// Get a message to show the current game result
+    let getMessage model = 
+        match getGameResult model with 
+        | StillPlaying -> sprintf "%O's turn" model.NextUp
         | XWins -> "X wins!"
         | OWins -> "O Wins!"
         | Draw -> "It is a draw!"
 
-    let announce m =
-        App.Current.MainPage.DisplayAlert("Game over", m, "OK") |> ignore
+    /// The 'update' function to update the model
+    let update gameOver msg model =
+        let newModel = 
+            match msg with
+            | Play pos -> { model with Board = model.Board.Add(pos, Full model.NextUp); NextUp = model.NextUp.Swap }
+            | Restart -> init()
 
-    let imageForPlayer p =
-        match p with
+        // Make an announcement in the middle of the game. 
+        let result = getGameResult newModel
+        if result <> StillPlaying then gameOver (getMessage newModel)
+
+        // Return the new model.
+        newModel
+
+    /// A helper used in the 'view' function to get the name 
+    /// of the Xaml resource for the image for a player
+    let imageForPlayer player =
+        match player with
         | X -> "Cross"
         | O -> "Nought"
 
-    let update msg m =
-        let model = match msg with
-                    | PlayTL -> { m with Cells = {m.Cells with TopLeft = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayTC -> { m with Cells = {m.Cells with TopCenter = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayTR -> { m with Cells = {m.Cells with TopRight = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayML -> { m with Cells = {m.Cells with MiddleLeft = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayMC -> { m with Cells = {m.Cells with MiddleCenter = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayMR -> { m with Cells = {m.Cells with MiddleRight = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayBL -> { m with Cells = {m.Cells with BottomLeft = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayBC -> { m with Cells = {m.Cells with BottomCenter = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | PlayBR -> { m with Cells = {m.Cells with BottomRight = Full m.Turn }; Turn = Player.Swap m.Turn }
-                    | Restart -> init()
-        let result = getGameResult model
-        if (result <> StillPlaying) then announce (getMessage model)
-        model
-
-    let canPlay m c =
-         match c with 
+    /// A helper to get the suffix used in the Xaml for a position on the board.
+    let uiText (row,col) = 
+        (match row with 0 -> "T" | 1 -> "M" | 2 -> "B" | _ -> failwith "huh?") + 
+        (match col with 0 -> "L" | 1 -> "C" | 2 -> "R" | _ -> failwith "huh?")
```

Comment:
It's pity we loose safety here :( Yes, it's boiler-free, but...

Reply:
yep, will adjust